### ISSUE TITLE
Implement face recognition

### ISSUE THREAD

**Issue by cirosantilli** [wavexx/facedetect#8]
I know nothing about face recognition, and this might be out of scope, but I suppose we could do something like this:

- extract faces from the images (using some python image library)
- encode / compress faces using some popular method
- store those on a sqlite database (Python has official support https://docs.python.org/2/library/sqlite3.html ) that maps encoded faces to image / rectangle pairs
- given a new image, encode it, and compare it to the ones on the database, and gives a list of the closest matches

Would be so awesome `*-*`

**Comment by wavexx** [comment]
I added recently some simple support to search for a single face through the new flag `-s`/`--search`, and although it's *far* from being optimal, it fits the concept of batch searching.

The idea is to supply one (or more - in the future) faces with `-s`, and scan an existing directory for similar faces.

I couldn't find an OSS face segmentation library, unfortunately. Using OpenCV the current best approach is to train SVM models to recognize a face, but this implies that you need many faces of the person you're trying to match, along with an equally big set of non-matching faces. I gave a few tries, and it works *much* better than what I do currently, but it's not what, as an user, I expect. I really don't want to train models to search for a face.

There might be situations where you definitely *could* build SVM models easily, but in this case you're better of by writing some dedicated code. For instance, corporations like Google or Facebook have the users providing them the pre-cut and tagged faces at zero cost, and there it's trivial to build accurate models to search for a single face.

For unsupervised search, you need a different approach, like face segmentation (ie: reconstructing some sort of ID that is scale/color/orientation invariant and thus suited to search).

I briefly tried to construct such a system using the HAAR features already available in OpenCV, but they came too short. The available models are poorly trained (left/right eyes, to name a few), and the precision of the feature matching is too low to construct anything on top. A bummer really.

I tried then to match two faces against identical SIFT features, but again, the problem here is that different features are picked according to the orientation of the face, and thus only a very marginal amount actually match even in the best scenario. The resolution of the scale was, again, too low to be useful.

I then used a very simple image-matching approach using MSSIM, and while not striking, it scores a good 30% on the [LFW](http://vis-www.cs.umass.edu/lfw/) dataset using the current parameters.

But if somebody has some pointer of OSS libraries that can help, I'm all ears.

**Comment by cirosantilli** [comment]
Ah, so you can't recognize faces well from a single image :-) I told you I was a newb. That is not a meaningful use case then.

By unsupervised search, you mean clustering faces from the same person without prior labels? Yes, that would be more interesting instead than what I had proposed.

**Comment by wavexx** [comment]
Exactly. We use HAAR to detect the face, then match the untagged face against all others, blindly. This is undoubtedly harder, but it's somewhat what *I* would expect as an user.
